src/ILCompiler.Compiler/src/Compiler/InteropStubManager.cs

```diff
@@ -101,32 +103,165 @@ internal TypeDesc GetInlineArrayType(InlineArrayCandidate candidate)
         }
 
         internal IEnumerable<DelegateMarshallingThunks> GetDelegateMarshallingThunks()
         {
             foreach (var delegateType in _delegateMarshalingTypes)
             {
-                var openStub = InteropStateManager.GetOpenStaticDelegateMarshallingThunk(delegateType);
-                var closedStub = InteropStateManager.GetClosedDelegateMarshallingThunk(delegateType);
-                var delegateCreationStub = InteropStateManager.GetForwardDelegateCreationThunk(delegateType);
                 yield return 
                     new DelegateMarshallingThunks()
                     {
                         DelegateType = delegateType,
-                        OpenStaticDelegateMarshallingThunk = openStub,
-                        ClosedDelegateMarshallingThunk = closedStub,
-                        DelegateCreationThunk = delegateCreationStub
+                        OpenStaticDelegateMarshallingThunk = InteropStateManager.GetOpenStaticDelegateMarshallingThunk(delegateType),
+                        ClosedDelegateMarshallingThunk = InteropStateManager.GetClosedDelegateMarshallingThunk(delegateType),
+                        DelegateCreationThunk = InteropStateManager.GetForwardDelegateCreationThunk(delegateType)
                     };
             }
         }
 
-        internal HashSet<NativeStructType> GetStructMarshallingTypes()
+        internal struct StructMarshallingThunks
+        {
+            public TypeDesc StructType;
+            public NativeStructType NativeStructType;
+            public MethodDesc MarshallingThunk;
+            public MethodDesc UnmarshallingThunk;
+            public MethodDesc CleanupThunk;
+        }
+
+        internal IEnumerable<StructMarshallingThunks> GetStructMarshallingTypes()
+        {
+                foreach (var structType in _structMarshallingTypes)
+            {
+                yield return
+                    new StructMarshallingThunks()
+                    {
+                        StructType = structType,
+                        NativeStructType = InteropStateManager.GetStructMarshallingNativeType(structType),
+                        MarshallingThunk = InteropStateManager.GetStructMarshallingManagedToNativeThunk(structType),
+                        UnmarshallingThunk = InteropStateManager.GetStructMarshallingNativeToManagedThunk(structType),
+                        CleanupThunk = InteropStateManager.GetStructMarshallingCleanupThunk(structType)
+                    };
+            }
+        }
+
+        public void AddDependeciesDueToPInvoke(ref DependencyList dependencies, NodeFactory factory, MethodDesc method)
+        {
+            if (method.IsPInvoke)
+            {
+                dependencies = dependencies ?? new DependencyList();
+
+                MethodSignature methodSig = method.Signature;
+                AddDependenciesDueToPInvokeDelegate(ref dependencies, factory, methodSig.ReturnType);
```

Comment:
It'd be good to add a wrapper called AddDepenendenciesForPInvokeType() which calls AddDependenciesDueToPinvokeDelegate/Struct/.... This way if you add a new one you don't have to add a call to all such places.